src/credssp/sspi_cred_ssp/tls_connection.rs

```diff
@@ -63,21 +136,152 @@ impl TlsConnection {
+            .ok_or_else(|| Error::new(ErrorKind::InternalError, "Can not query negotiated TLS version"))?
+            .get_u16()
+            .to_be_bytes();
+        tls_packet_start.extend_from_slice(&tls_version);
+
+        // Safe: payload length is checked above.
+        if payload[0..1 /* ContentType */ + 2 /* ProtocolVersion */] != tls_packet_start {
+            return Err(Error::new(ErrorKind::InvalidToken, "Invalid TLS packet header."));
+        }
+
+        // Safe: payload length is checked above.
+        let encrypted_application_data_len = usize::from(u16::from_be_bytes(payload[3..5].try_into().unwrap()));
+
+        if payload.len() < TLS_PACKET_HEADER_LEN + encrypted_application_data_len {
+            return Err(Error::new(ErrorKind::InvalidToken, "Input TLS buffer is too short."));
+        }
+
+        // Safe: payload length is checked above.
+        let (header, rest) = payload.split_at_mut(TLS_PACKET_PREFIX_LEN);
+        // `encrypted_application_data_len` is a len of the encrypted data with the sequence number.
+        // But here we need the encrypted data *WITHOUT* a sequence number, so we subtract TLS_PACKET_SEQUENCE_NUMBER_LEN
+        // from the overall data length.
+        let (application_data, extra) =
+            rest.split_at_mut(encrypted_application_data_len - TLS_PACKET_SEQUENCE_NUMBER_LEN);
+
+        Ok(TlsTrafficParts {
+            header,
+            application_data,
+            extra,
+        })
+    }
+
+    /// Decrypt a part of the incoming TLS traffic.
+    ///
+    /// If the input buffer contains more than one TLS message,then only the first one will be decrypted.
+    pub fn decrypt_tls<'a>(&mut self, payload: &'a mut [u8]) -> Result<DecryptionResult<'a>> {
         match self {
             TlsConnection::Rustls(tls_connection) => {
-                tls_connection.read_tls(&mut payload)?;
+                let mut tls_packet = match TlsConnection::find_tls_data_to_decrypt(tls_connection, payload)? {
+                    FindTlsPacketResult::TlsPacket(data) => data as &[u8],
+                    FindTlsPacketResult::Missing(needed_bytes_amount) => {
+                        return Ok(DecryptionResult::IncompleteMessage(needed_bytes_amount));
+                    }
+                };
+                let mut plain_data = Vec::with_capacity(tls_packet.len());
 
-                let tls_state = tls_connection
-                    .process_new_packets()
-                    .map_err(|err| Error::new(ErrorKind::DecryptFailure, err.to_string()))?;
+                while !tls_packet.is_empty() {
+                    let _ = tls_connection.read_tls(&mut tls_packet)?;
 
-                let mut reader = tls_connection.reader();
+                    let tls_state = tls_connection
+                        .process_new_packets()
+                        .map_err(|err| Error::new(ErrorKind::DecryptFailure, err.to_string()))?;
+                    trace!(?tls_state);
```

Comment:
nope, it's not a leftover 😃 
It's very convenient to see how many bytes the `rustls` has managed to decrypt. Also, it shows us an overall IO state.
Sometimes, it can be very useful during debugging.

Reply:
done: https://github.com/Devolutions/sspi-rs/issues/217 

I'll link to this issue example of the solution for the decryption problem once the corresponding PR is merged.